### ISSUE TITLE
Improve Batch training

### ISSUE THREAD

**Issue by nateanl** [speechLabBcCuny/onssen#5]
The current training code calls the backward function at every batch. Theoretically, the function is called only once an epoch, it saves the loss for each batch and backward at the end. This may cause the memory explosion.
